CORGI-200 compose modules

jasinner opened the pull request:
This builds the component tree to link RHEL_MODULE component with it's RPM component children when ingesting composes.

This also fixes CORGI-211

jasinner committed:
fetch module build_id for save to relation

jasinner committed:
fetch and save modules to core component taxonomy models

jasinner committed:
make sure parent is part of cnode null true index

mprpic commented on corgi/collectors/rhel_compose.py:
```diff
@@ -64,22 +102,97 @@ def fetch_compose_data(
+        brew = Brew()
+        response = requests.get(compose_url + "rpms.json")
+        if response.ok:
+            rpm_filnames_by_srpm = defaultdict(list)
```
```suggestion
            rpm_filenames_by_srpm = defaultdict(list)
```
?

jasinner replied:
Done.

jasinner replied:
Done.

mprpic commented on corgi/collectors/models.py:
```diff
@@ -31,3 +31,32 @@ class CollectorErrataProductVariant(CollectorErrataModel):
+
+
+# Not every RPM built as part of a SRPM is included in a module
+# Which is why we don't directly related SRPM builds to product_streams using relations table
```
Can you make this a doc string of the class below?

mprpic reviewed (approved):
Looks good, just minor nits.

The relations stuff is a bit of a black box for me but I assume it's use to relate a compose to a product and a set of builds? I.e. an alternative to errata?

jasinner commented on corgi/collectors/rhel_compose.py:
```diff
@@ -2,22 +2,73 @@
+        try:
+            rhel_module = CollectorComposeRhelModule.objects.get(build_id=build_id)
+        except CollectorComposeRhelModule.DoesNotExist:
+            logger.debug("Did not find %s in CollectorComposeRhelModule data", build_id)
```
It's expected behavior that CCRhelModule does not exists in many cases. That's because we also have many non-modular SRPMs in COMPOSE relations as well. All of those will log something, which is why I set it to DEBUG, because we probably don't care 99% of the time.

jasinner committed:
include release in rhel_module purl

jasinner commented:
I made 3 minor change since the review. I fixed an issue where release was not included as part of the RHEL_MODULE component purl. I also updated the original save_module function of the brew task to use update_or_create instead of get_or_create so that the software_build foreign key was set on the RHEL_MODULE even if it was created by the fetch_compose_build task previously. And finally I adjust the software_composition_analysis task so that it looked up RHEL_MODULE as well as SRPMs so it could link the discovered components to the correct root node, and didn't through an error about the Component not existing.

I verified that both SRPM and RHEL_MODULE build_ids are added to relations table for loadcomposedata commands to later lookup and call fetch_brew_build on.

Here's a few queries that show the resulting data layout:

```
for rm in m.Component.objects.filter(type="RHEL_MODULE"):
     rm.save_component_taxonomy()
     print(f"RHEL_MODULE:{rm.purl}")
     for provide in rm.provides:
         print(provide)
     print("-----------------")
```
Shows the following Component are related:
```
-----------------
RHEL_MODULE:pkg:rhel_module/redhat/virt-devel@rhel-8060020220408104655.d63f516d
pkg:rpm/redhat/libvirt-dbus-debugsource@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=i686
```

Checking the sources of the RPM, we see both it's SRPM parent and it's RHEL_MODULE:
```
rpm = m.Component.objects.get(
     purl="pkg:rpm/redhat/libvirt-dbus-debugsource@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=i686"
)
rpm.save_component_taxonomy()
rpm.sources
['pkg:srpm/redhat/libvirt-dbus@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=src',
 'pkg:rhel_module/redhat/virt-devel@rhel-8060020220408104655.d63f516d']
```

Using the Collector models to check the rpm's source:
```
cm.CollectorComposeRPM.objects.get(nvra=f"{rpm.nvr}.{rpm.arch}").srpm
<CollectorComposeSRPM: 1929918>
```

Verifying that that RHEL_MODULE component does *not* list the 1929918 SoftwareBuild as a foreign key:
```
for c in m.Component.objects.filter(software_build__build_id=1929918):
    ...:     print(c.purl)
pkg:rpm/redhat/libvirt-dbus@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=x86_64
pkg:rpm/redhat/libvirt-dbus@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=i686
pkg:rpm/redhat/libvirt-dbus@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=ppc64le
pkg:rpm/redhat/libvirt-dbus@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=aarch64
pkg:rpm/redhat/libvirt-dbus@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=s390x
pkg:rpm/redhat/libvirt-dbus-debuginfo@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=s390x
pkg:rpm/redhat/libvirt-dbus-debuginfo@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=i686
pkg:rpm/redhat/libvirt-dbus-debuginfo@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=x86_64
pkg:rpm/redhat/libvirt-dbus-debuginfo@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=aarch64
pkg:rpm/redhat/libvirt-dbus-debuginfo@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=ppc64le
pkg:rpm/redhat/libvirt-dbus-debugsource@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=i686
pkg:rpm/redhat/libvirt-dbus-debugsource@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=aarch64
pkg:rpm/redhat/libvirt-dbus-debugsource@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=ppc64le
pkg:rpm/redhat/libvirt-dbus-debugsource@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=s390x
pkg:rpm/redhat/libvirt-dbus-debugsource@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=x86_64
pkg:srpm/redhat/libvirt-dbus@1.3.0-2.module%2Bel8.6.0%2B14480%2Bc0a3aa0f?arch=src
```

Confirming that the RHEL_MODULE is related to it's SoftwareBuild:
```
cm.CollectorComposeRhelModule.objects.get(
    ...:     nvr="virt-devel-rhel-8060020220408104655.d63f516d"
    ...: ).build_id
1964073
```

Looking at the RHEL_MODULE and checking it's softwarebuild foreign key:
```
module = m.Component.objects.get(
     purl="pkg:rhel_module/redhat/virt-devel@rhel-8060020220408104655.d63f516d"
)
module.software_build
<SoftwareBuild: SoftwareBuild object (1964073)>
```

jasinner committed:
use update_or_create instead of get_or_create for rhel_modules in brew task to ensure software_build is added to the RHEL_MODULE component

jasinner committed:
adjust sca task to scan sources for RHEL_MODULE component types also
